File: files/Q7_Medium_ResourceLeak.java
/**
 * PROBLEM: Configuration File Reader
 * 
 * Read configuration settings from a file and parse them.
 * The program should handle file operations safely and close resources properly.
 * 
 * File format (config.txt):
 * username=admin
 * timeout=30
 * max_connections=100
 * 
 * INPUT: Filename
 * OUTPUT: Parsed configuration as key-value pairs
 * 
 * EXPECTED BEHAVIOR:
 * - Read file successfully
 * - Parse all key-value pairs
 * - Close file properly even if errors occur
 * - Handle missing files gracefully
 * 
 * BUG: The code has a resource management issue. Find and fix the bug!
 */

import java.io.*;
import java.util.*;

public class Q7_Medium_ResourceLeak {
    
    public static Map<String, String> readConfig(String filename) {
        Map<String, String> config = new HashMap<>();
        BufferedReader reader = null;
        
        try {
            reader = new BufferedReader(new FileReader(filename));
            String line;
            
            while ((line = reader.readLine()) != null) {
                line = line.trim();
                
                // Skip empty lines and comments
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }
                
                // Parse key=value pairs
                if (line.contains("=")) {
                    String[] parts = line.split("=", 2);
                    String key = parts[0].trim();
                    String value = parts[1].trim();
                    config.put(key, value);
                } else {
                    throw new IllegalArgumentException("Invalid format: " + line);
                }
            }
            
            reader.close();
            
        } catch (FileNotFoundException e) {
            System.out.println("Error: File not found - " + filename);
            return null;
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return null;
        } catch (IllegalArgumentException e) {
            System.out.println("Error parsing config: " + e.getMessage());
            return null;
            // BUG HERE: Reader is not closed when exception occurs
            // Should use try-with-resources or finally block
        }
        
        return config;
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        System.out.println("Enter config filename:");
        String filename = sc.nextLine();
        
        Map<String, String> config = readConfig(filename);
        
        if (config != null) {
            System.out.println("\nConfiguration loaded:");
            for (Map.Entry<String, String> entry : config.entrySet()) {
                System.out.println(entry.getKey() + " = " + entry.getValue());
            }
        }
        
        sc.close();
    }
}

File: files/Q7_Medium_ResourceLeak.py
"""
PROBLEM: Configuration File Reader

Read configuration settings from a file and parse them.
The program should handle file operations safely and close resources properly.

File format (config.txt):
username=admin
timeout=30
max_connections=100

INPUT: Filename
OUTPUT: Parsed configuration as key-value pairs

EXPECTED BEHAVIOR:
- Read file successfully
- Parse all key-value pairs
- Close file properly even if errors occur
- Handle missing files gracefully

BUG: The code has a resource management issue. Find and fix the bug!
"""

def readConfig(filename):
    config = {}
    
    try:
        file = open(filename, 'r')
        
        for line in file:
            line = line.strip()
            
            # Skip empty lines and comments
            if not line or line.startswith('#'):
                continue
            
            # Parse key=value pairs
            if '=' in line:
                parts = line.split('=', 1)
                key = parts[0].strip()
                value = parts[1].strip()
                config[key] = value
            else:
                print(f"Error parsing config: Invalid format - {line}")
                return None
                # BUG HERE: File is not closed when error occurs
                # Should use 'with' statement or finally block
        
        file.close()
        
    except FileNotFoundError:
        print(f"Error: File not found - {filename}")
        return None
    except Exception as e:
        print(f"Error reading file: {e}")
        return None
        # BUG: File is not closed in exception cases
    
    return config

def main():
    filename = input("Enter config filename:\n")
    
    config = readConfig(filename)
    
    if config is not None:
        print("\nConfiguration loaded:")
        for key, value in config.items():
            print(f"{key} = {value}")

if __name__ == "__main__":
    main()
